knative/deployer.go

```diff
@@ -80,7 +110,41 @@ func (d *Deployer) Deploy(ctx context.Context, f fn.Function) (result fn.Deploym
 			if d.Verbose {
 				fmt.Println("Waiting for Knative Service to become ready")
 			}
-			err, _ = client.WaitForService(ctx, f.Name, DefaultWaitingTimeout, wait.NoopMessageCallback())
+			chprivate := make(chan bool)
+			cherr := make(chan error)
+			go func() {
+				private := false
+				for !private {
+					time.Sleep(5 * time.Second)
+					private = d.isImageInPrivateRegistry(ctx, client, f.Name)
+					chprivate <- private
+				}
+				close(chprivate)
+			}()
+			go func() {
+				err, _ := client.WaitForService(ctx, f.Name, DefaultWaitingTimeout, wait.NoopMessageCallback())
+				cherr <- err
+				close(cherr)
+			}()
+
+			presumePrivate := false
+		main:
+			// Wait for either a timeout or a container condition signaling the image is unreachable
+			for {
+				select {
+				case private := <-chprivate:
+					if private {
+						presumePrivate = true
+						break main
+					}
+				case err = <-cherr:
+					break main
+				}
+			}
+			if presumePrivate {
+				err := fmt.Errorf("it seems your docker registry is private, make sure you have set up pull secrets https://knative.dev/docs/developer/serving/deploying-from-private-registry")
```

Comment:
I really like this PR, but I'm not sure if we should say this so concretely. It's really only an assumption on our part. Would something like this be better?
```suggestion
				err := fmt.Errorf("your function image is unreachable. It is possible that your docker registry is private. If so, make sure you have set up pull secrets https://knative.dev/docs/developer/serving/deploying-from-private-registry")
```